tracer.go

```diff
@@ -52,38 +76,54 @@ func (t *Tracer) StartSpan(
 		option(t, s)
 	}
 
+	if (model.SpanContext{}) != s.SpanContext {
+		// we received a parent SpanContext
+		if t.sharedSpans && s.Kind == model.Server {
+			// join span
+			s.Shared = true
+		} else {
+			// regular child span
+			parentID := s.ID
+			s.ParentID = &parentID
+			s.ID = t.generate.SpanID(model.TraceID{})
+		}
+	}
+
 	// test if extraction resulted in an error
-	if s.SpanContext.err != nil {
-		switch t.options.extractFailurePolicy {
+	if s.SpanContext.Err != nil {
+		switch t.extractFailurePolicy {
 		case ExtractFailurePolicyRestart:
 		case ExtractFailurePolicyError:
-			panic(s.SpanContext.err)
+			panic(s.SpanContext.Err)
 		case ExtractFailurePolicyTagAndRestart:
-			s.Tags["error.extract"] = s.SpanContext.err.Error()
+			s.Tags["error.extract"] = s.SpanContext.Err.Error()
 		default:
 			panic(ErrInvalidExtractFailurePolicy)
 		}
 		// restart the trace
-		s.SpanContext.TraceID = t.options.generate.TraceID()
-		s.SpanContext.ID = t.options.generate.SpanID()
+		s.SpanContext.TraceID = t.generate.TraceID()
+		s.SpanContext.ID = t.generate.SpanID(s.SpanContext.TraceID)
 		s.SpanContext.ParentID = nil
-		s.SpanContext.err = nil
 	} else if s.SpanContext.TraceID.Empty() || s.SpanContext.ID == 0 {
 		// create root span
-		s.SpanContext.TraceID = t.options.generate.TraceID()
-		s.SpanContext.ID = t.options.generate.SpanID()
+		s.SpanContext.TraceID = t.generate.TraceID()
+		s.SpanContext.ID = t.generate.SpanID(s.SpanContext.TraceID)
 	}
 
 	if !s.SpanContext.Debug && s.Sampled == nil {
```

Comment:
I'd really move all this logic into the context itself, it feels like we know too much about the context rather than the context itself.

Reply:
SpanContext and other items in the model package are purely a storage and json serialization/deserialization effort. By decoupling from the tracer implementation they can be easily reused for other efforts to integrate with Zipkin. I don't want to clutter them with business logic which might not map / align with those efforts for being to zipkin-go library specific.